File: python/langchain_kserve/_openai_compat.py
"""OpenAI-compatible endpoint request/response mapping for KServe.

Handles ``POST /v1/chat/completions`` and ``POST /v1/completions`` as served
by runtimes like vLLM and TGI.
"""

from __future__ import annotations

import json
import logging
from typing import Any, AsyncIterator, Dict, Iterator, List, Optional, Sequence

import httpx
from langchain_core.messages import (
    AIMessage,
    AIMessageChunk,
    BaseMessage,
    HumanMessage,
    SystemMessage,
    ToolMessage,
)
from langchain_core.messages.tool import ToolCall
from langchain_core.outputs import ChatGeneration, ChatGenerationChunk, ChatResult

from langchain_kserve._common import (
    KServeInferenceError,
    async_request_with_retry,
    request_with_retry,
)

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Message serialisation
# ---------------------------------------------------------------------------


def messages_to_openai_dicts(messages: Sequence[BaseMessage]) -> List[Dict[str, Any]]:
    """Convert LangChain messages to the OpenAI ``messages`` list format.

    Args:
        messages: Sequence of :class:`~langchain_core.messages.BaseMessage`.

    Returns:
        List of dicts suitable for the ``messages`` field in a chat completion
        request body.
    """
    result: List[Dict[str, Any]] = []
    for msg in messages:
        if isinstance(msg, SystemMessage):
            result.append({"role": "system", "content": _get_text_content(msg)})
        elif isinstance(msg, HumanMessage):
            result.append({"role": "user", "content": _get_text_content(msg)})
        elif isinstance(msg, AIMessage):
            d: Dict[str, Any] = {"role": "assistant", "content": msg.content or ""}
            if msg.tool_calls:
                d["tool_calls"] = [
                    {
                        "id": tc["id"],
                        "type": "function",
                        "function": {
                            "name": tc["name"],
                            "arguments": json.dumps(tc["args"]),
                        },
                    }
                    for tc in msg.tool_calls
                ]
            result.append(d)
        elif isinstance(msg, ToolMessage):
            result.append(
                {
                    "role": "tool",
                    "tool_call_id": msg.tool_call_id,
                    "content": str(msg.content),
                }
            )
        else:
            # Fallback: treat as user message
            result.append({"role": "user", "content": _get_text_content(msg)})
    return result


def _get_text_content(msg: BaseMessage) -> str:
    if isinstance(msg.content, str):
        return msg.content
    # Handle list of content blocks (multimodal)
    parts: List[str] = []
    for block in msg.content:  # type: ignore[union-attr]
        if isinstance(block, str):
            parts.append(block)
        elif isinstance(block, dict) and block.get("type") == "text":
            parts.append(block.get("text", ""))
    return "".join(parts)


# ---------------------------------------------------------------------------
# Request builders
# ---------------------------------------------------------------------------


def build_chat_request(
    model_name: str,
    messages: Sequence[BaseMessage],
    temperature: float,
    max_tokens: Optional[int],
    top_p: float,
    stop: Optional[List[str]],
    stream: bool,
    tools: Optional[List[Dict[str, Any]]] = None,
    extra_kwargs: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Build a ``/v1/chat/completions`` request body.

    Args:
        model_name: The model identifier.
        messages: Conversation messages.
        temperature: Sampling temperature.
        max_tokens: Max tokens to generate (omitted if ``None``).
        top_p: Nucleus sampling parameter.
        stop: Stop sequences.
        stream: Whether to stream the response.
        tools: OpenAI-format tool schemas (for tool calling).
        extra_kwargs: Additional parameters merged into the request body.

    Returns:
        JSON-serialisable request body dict.
    """
    body: Dict[str, Any] = {
        "model": model_name,
        "messages": messages_to_openai_dicts(messages),
        "temperature": temperature,
        "top_p": top_p,
        "stream": stream,
    }
    if max_tokens is not None:
        body["max_tokens"] = max_tokens
    if stop:
        body["stop"] = stop
    if tools:
        body["tools"] = tools
    if extra_kwargs:
        body.update(extra_kwargs)
    return body


def build_completion_request(
    model_name: str,
    prompt: str,
    temperature: float,
    max_tokens: Optional[int],
    top_p: float,
    stop: Optional[List[str]],
    stream: bool,
    extra_kwargs: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Build a ``/v1/completions`` request body.

    Args:
        model_name: The model identifier.
        prompt: The text prompt.
        temperature: Sampling temperature.
        max_tokens: Max tokens to generate.
        top_p: Nucleus sampling.
        stop: Stop sequences.
        stream: Whether to stream.
        extra_kwargs: Extra parameters merged into the body.

    Returns:
        JSON-serialisable request body dict.
    """
    body: Dict[str, Any] = {
        "model": model_name,
        "prompt": prompt,
        "temperature": temperature,
        "top_p": top_p,
        "stream": stream,
    }
    if max_tokens is not None:
        body["max_tokens"] = max_tokens
    if stop:
        body["stop"] = stop
    if extra_kwargs:
        body.update(extra_kwargs)
    return body


# ---------------------------------------------------------------------------
# Response parsers
# ---------------------------------------------------------------------------


def parse_chat_response(
    response: httpx.Response,
    model_name: str,
    protocol_used: str = "openai",
) -> ChatResult:
    """Parse a non-streaming ``/v1/chat/completions`` response.

    Args:
        response: The raw HTTP response.
        model_name: Model name for generation metadata.
        protocol_used: Protocol label for generation info.

    Returns:
        A :class:`~langchain_core.outputs.ChatResult`.

    Raises:
        KServeInferenceError: If the response format is unexpected.
    """
    try:
        data = response.json()
    except Exception as exc:
        raise KServeInferenceError(f"Failed to parse response JSON: {exc}") from exc

    choices = data.get("choices", [])
    if not choices:
        raise KServeInferenceError(f"No choices in response: {data}")

    choice = choices[0]
    message = choice.get("message", {})
    content: str = message.get("content") or ""
    finish_reason: Optional[str] = choice.get("finish_reason")

    tool_calls: List[ToolCall] = []
    raw_tool_calls = message.get("tool_calls") or []
    for tc in raw_tool_calls:
        fn = tc.get("function", {})
        try:
            args = json.loads(fn.get("arguments", "{}"))
        except json.JSONDecodeError:
            args = {}
        tool_calls.append(
            ToolCall(
                id=tc.get("id", ""),
                name=fn.get("name", ""),
                args=args,
            )
        )

    ai_message = AIMessage(content=content, tool_calls=tool_calls)

    usage = data.get("usage", {})
    generation_info: Dict[str, Any] = {
        "model": model_name,
        "protocol": protocol_used,
        "finish_reason": finish_reason,
        "prompt_tokens": usage.get("prompt_tokens"),
        "completion_tokens": usage.get("completion_tokens"),
        "total_tokens": usage.get("total_tokens"),
    }

    return ChatResult(
        generations=[ChatGeneration(message=ai_message, generation_info=generation_info)]
    )


def parse_completion_response(response: httpx.Response, model_name: str) -> str:
    """Parse a non-streaming ``/v1/completions`` response.

    Args:
        response: The raw HTTP response.
        model_name: Model name (for logging).

    Returns:
        The generated text string.

    Raises:
        KServeInferenceError: If the response format is unexpected.
    """
    try:
        data = response.json()
    except Exception as exc:
        raise KServeInferenceError(f"Failed to parse response JSON: {exc}") from exc

    choices = data.get("choices", [])
    if not choices:
        raise KServeInferenceError(f"No choices in response: {data}")

    return choices[0].get("text", "")


# ---------------------------------------------------------------------------
# Sync streaming
# ---------------------------------------------------------------------------


def stream_chat_response(
    client: httpx.Client,
    path: str,
    body: Dict[str, Any],
    max_retries: int,
    model_name: str,
    protocol_used: str = "openai",
) -> Iterator[ChatGenerationChunk]:
    """Stream ``/v1/chat/completions`` and yield :class:`ChatGenerationChunk`.

    Args:
        client: Sync HTTP client.
        path: Endpoint path (``"/v1/chat/completions"``).
        body: Request body (must have ``stream=True``).
        max_retries: Retry count for the initial connection.
        model_name: Model name for generation metadata.
        protocol_used: Protocol label.

    Yields:
        :class:`~langchain_core.outputs.ChatGenerationChunk` for each SSE token.
    """
    with client.stream("POST", path, json=body) as response:
        _raise_stream_status(response)
        for line in response.iter_lines():
            chunk = _parse_sse_chat_line(line, model_name, protocol_used)
            if chunk is not None:
                yield chunk


def stream_completion_response(
    client: httpx.Client,
    path: str,
    body: Dict[str, Any],
    model_name: str,
) -> Iterator[str]:
    """Stream ``/v1/completions`` and yield text deltas.

    Args:
        client: Sync HTTP client.
        path: Endpoint path.
        body: Request body.
        model_name: Model name.

    Yields:
        Text delta strings.
    """
    with client.stream("POST", path, json=body) as response:
        _raise_stream_status(response)
        for line in response.iter_lines():
            text = _parse_sse_completion_line(line)
            if text is not None:
                yield text


# ---------------------------------------------------------------------------
# Async streaming
# ---------------------------------------------------------------------------


async def astream_chat_response(
    client: httpx.AsyncClient,
    path: str,
    body: Dict[str, Any],
    model_name: str,
    protocol_used: str = "openai",
) -> AsyncIterator[ChatGenerationChunk]:
    """Async version of :func:`stream_chat_response`.

    Args:
        client: Async HTTP client.
        path: Endpoint path.
        body: Request body (must have ``stream=True``).
        model_name: Model name.
        protocol_used: Protocol label.

    Yields:
        :class:`~langchain_core.outputs.ChatGenerationChunk` per SSE token.
    """
    async with client.stream("POST", path, json=body) as response:
        _raise_stream_status(response)
        async for line in response.aiter_lines():
            chunk = _parse_sse_chat_line(line, model_name, protocol_used)
            if chunk is not None:
                yield chunk


async def astream_completion_response(
    client: httpx.AsyncClient,
    path: str,
    body: Dict[str, Any],
    model_name: str,
) -> AsyncIterator[str]:
    """Async version of :func:`stream_completion_response`.

    Yields:
        Text delta strings.
    """
    async with client.stream("POST", path, json=body) as response:
        _raise_stream_status(response)
        async for line in response.aiter_lines():
            text = _parse_sse_completion_line(line)
            if text is not None:
                yield text


# ---------------------------------------------------------------------------
# SSE parsing helpers
# ---------------------------------------------------------------------------


def _parse_sse_chat_line(
    line: str,
    model_name: str,
    protocol_used: str,
) -> Optional[ChatGenerationChunk]:
    line = line.strip()
    if not line or not line.startswith("data:"):
        return None
    payload = line[len("data:"):].strip()
    if payload == "[DONE]":
        return None
    try:
        data = json.loads(payload)
    except json.JSONDecodeError:
        return None

    choices = data.get("choices", [])
    if not choices:
        return None
    delta = choices[0].get("delta", {})
    content: str = delta.get("content") or ""
    finish_reason = choices[0].get("finish_reason")

    # Handle tool call deltas
    tool_call_chunks = delta.get("tool_calls")

    generation_info: Dict[str, Any] = {
        "model": model_name,
        "protocol": protocol_used,
        "finish_reason": finish_reason,
    }

    return ChatGenerationChunk(
        message=AIMessageChunk(
            content=content,
            tool_call_chunks=tool_call_chunks or [],
        ),
        generation_info=generation_info,
    )


def _parse_sse_completion_line(line: str) -> Optional[str]:
    line = line.strip()
    if not line or not line.startswith("data:"):
        return None
    payload = line[len("data:"):].strip()
    if payload == "[DONE]":
        return None
    try:
        data = json.loads(payload)
    except json.JSONDecodeError:
        return None
    choices = data.get("choices", [])
    if not choices:
        return None
    return choices[0].get("text") or ""


def _raise_stream_status(response: httpx.Response) -> None:
    if not response.is_success:
        raise KServeInferenceError(
            f"KServe stream request failed ({response.status_code})"
        )


# ---------------------------------------------------------------------------
# Protocol detection
# ---------------------------------------------------------------------------


def probe_openai_compat(
    client: httpx.Client,
    max_retries: int,
) -> bool:
    """Check if the endpoint exposes an OpenAI-compatible ``/v1/models`` route.

    Args:
        client: Sync HTTP client.
        max_retries: Retry count.

    Returns:
        ``True`` if the endpoint responds with HTTP 200.
    """
    try:
        response = request_with_retry(client, "GET", "/v1/models", max_retries)
        return response.status_code == 200
    except Exception:
        return False


async def async_probe_openai_compat(
    client: httpx.AsyncClient,
    max_retries: int,
) -> bool:
    """Async version of :func:`probe_openai_compat`.

    Args:
        client: Async HTTP client.
        max_retries: Retry count.

    Returns:
        ``True`` if the endpoint responds with HTTP 200.
    """
    try:
        response = await async_request_with_retry(
            client, "GET", "/v1/models", max_retries
        )
        return response.status_code == 200
    except Exception:
        return False

File: typescript/src/openai-compat.ts
/**
 * OpenAI-compatible endpoint request building and response parsing.
 *
 * Handles:
 * - /v1/chat/completions (non-streaming + streaming)
 * - /v1/completions (non-streaming + streaming)
 * - Tool call extraction from responses
 */

import type { BaseMessage } from "@langchain/core/messages";
import {
  AIMessage,
  AIMessageChunk,
} from "@langchain/core/messages";
import type { InvalidToolCall } from "@langchain/core/messages/tool";
import type { ChatGeneration, LLMResult } from "@langchain/core/outputs";
import { ChatGenerationChunk } from "@langchain/core/outputs";

import { KServeInferenceError } from "./errors.js";
import type {
  ChatKServeCallOptions,
  KServeGenerationInfo,
  OpenAIChatRequest,
  OpenAIChatResponse,
  OpenAIChatStreamChunk,
  OpenAICompletionRequest,
  OpenAICompletionResponse,
  OpenAICompletionStreamChunk,
  OpenAITool,
  OpenAIToolCall,
  OpenAIUsage,
} from "./types.js";
import { convertMessagesToOpenAI } from "./utils.js";

// ============================================================
// Chat completions
// ============================================================

/**
 * Build an OpenAI chat completions request body.
 *
 * @param modelName - Model identifier
 * @param messages  - LangChain messages to convert
 * @param params    - Generation parameters
 * @param tools     - Optional tool definitions
 * @param stream    - Whether to request streaming
 */
export function buildChatRequest(
  modelName: string,
  messages: BaseMessage[],
  params: {
    temperature?: number;
    maxTokens?: number;
    topP?: number;
    stop?: string[];
  },
  options?: Partial<ChatKServeCallOptions>,
  stream = false
): OpenAIChatRequest {
  const request: OpenAIChatRequest = {
    model: modelName,
    messages: convertMessagesToOpenAI(messages),
    stream,
  };

  const temperature = options?.temperature ?? params.temperature;
  const maxTokens = options?.maxTokens ?? params.maxTokens;

  if (temperature !== undefined) request.temperature = temperature;
  if (maxTokens !== undefined) request.max_tokens = maxTokens;
  if (params.topP !== undefined) request.top_p = params.topP;

  const stop = options?.stop ?? params.stop;
  if (stop && stop.length > 0) request.stop = stop;

  // Tool calling
  const tools: OpenAITool[] | undefined = options?.tools;
  if (tools && tools.length > 0) {
    request.tools = tools;
    if (options?.toolChoice !== undefined) {
      // Cast needed because our union type is wider than OpenAIChatRequest expects
      request.tool_choice = options.toolChoice as typeof request.tool_choice;
    }
    if (options?.parallelToolCalls !== undefined) {
      request.parallel_tool_calls = options.parallelToolCalls;
    }
  }

  // Logprobs
  if (options?.logprobs !== undefined) {
    request.logprobs = options.logprobs;
  }
  if (options?.topLogprobs !== undefined) {
    request.top_logprobs = options.topLogprobs;
  }

  // Request usage in streaming chunks (vLLM supports this)
  if (stream) {
    request.stream_options = { include_usage: true };
  }

  return request;
}

/**
 * Parse an OpenAI chat completions response into a LangChain ChatGeneration.
 *
 * @param response  - Raw API response
 * @param modelName - Model name for metadata
 */
export function parseChatResponse(
  response: OpenAIChatResponse,
  modelName: string
): ChatGeneration {
  const choice = response.choices[0];
  if (!choice) {
    throw new KServeInferenceError("OpenAI response contained no choices");
  }

  const msg = choice.message;
  const toolCalls = msg.tool_calls;

  const generationInfo: KServeGenerationInfo = {
    modelName: response.model ?? modelName,
    protocol: "openai",
    finishReason: choice.finish_reason,
  };

  if (response.usage) {
    generationInfo.tokenUsage = {
      promptTokens: response.usage.prompt_tokens,
      completionTokens: response.usage.completion_tokens,
      totalTokens: response.usage.total_tokens,
    };
  }

  generationInfo.logprobs = choice.logprobs ?? null;

  let aiMessage: AIMessage;

  if (toolCalls && toolCalls.length > 0) {
    const validToolCalls: Array<{
      id: string;
      name: string;
      args: Record<string, unknown>;
      type: "tool_call";
    }> = [];
    const invalidToolCalls: InvalidToolCall[] = [];

    for (const tc of toolCalls) {
      try {
        const parsed: unknown = JSON.parse(tc.function.arguments);
        let args: Record<string, unknown>;
        if (typeof parsed === "object" && parsed !== null) {
          args = parsed as Record<string, unknown>;
        } else {
          args = { input: parsed };
        }
        validToolCalls.push({
          id: tc.id,
          name: tc.function.name,
          args,
          type: "tool_call" as const,
        });
      } catch (e) {
        invalidToolCalls.push({
          id: tc.id,
          name: tc.function.name,
          args: tc.function.arguments,
          error: e instanceof Error ? e.message : String(e),
          type: "invalid_tool_call" as const,
        });
      }
    }

    aiMessage = new AIMessage({
      content: msg.content ?? "",
      tool_calls: validToolCalls,
      invalid_tool_calls: invalidToolCalls,
    });
  } else {
    aiMessage = new AIMessage({ content: msg.content ?? "" });
  }

  return {
    message: aiMessage,
    text: msg.content ?? "",
    generationInfo,
  };
}


/**
 * Parse a streaming OpenAI chat chunk and return a ChatGenerationChunk.
 * Returns null if the chunk carries no useful delta.
 */
export function parseChatStreamChunk(
  raw: string,
  modelName: string
): ChatGenerationChunk | null {
  let chunk: OpenAIChatStreamChunk;
  try {
    chunk = JSON.parse(raw) as OpenAIChatStreamChunk;
  } catch {
    return null;
  }

  const choice = chunk.choices?.[0];
  if (!choice) return null;

  const delta = choice.delta;
  const content = delta.content ?? "";
  const finishReason = choice.finish_reason ?? undefined;

  const generationInfo: KServeGenerationInfo & Record<string, unknown> = {
    modelName: chunk.model ?? modelName,
    protocol: "openai",
    finishReason,
  };

  if (chunk.usage) {
    generationInfo.tokenUsage = {
      promptTokens: chunk.usage.prompt_tokens,
      completionTokens: chunk.usage.completion_tokens,
      totalTokens: chunk.usage.total_tokens,
    };
  }

  // Logprobs (per-chunk, when requested)
  if (choice.logprobs !== undefined) {
    generationInfo.logprobs = choice.logprobs ?? null;
  }

  // Handle streaming tool calls
  if (delta.tool_calls && delta.tool_calls.length > 0) {
    const messageChunk = new AIMessageChunk({
      content,
      tool_call_chunks: delta.tool_calls.map((t) => ({
        index: t.index,
        id: t.id,
        name: t.function?.name,
        args: t.function?.arguments,
        type: "tool_call_chunk" as const,
      })),
    });
    return new ChatGenerationChunk({
      text: content,
      message: messageChunk,
      generationInfo,
    });
  }

  return new ChatGenerationChunk({
    text: content,
    message: new AIMessageChunk({ content }),
    generationInfo,
  });
}

// ============================================================
// Text completions
// ============================================================

/**
 * Build an OpenAI completions (non-chat) request body.
 */
export function buildCompletionRequest(
  modelName: string,
  prompt: string,
  params: {
    temperature?: number;
    maxTokens?: number;
    topP?: number;
    stop?: string[];
  },
  stream = false
): OpenAICompletionRequest {
  const request: OpenAICompletionRequest = {
    model: modelName,
    prompt,
    stream,
  };

  if (params.temperature !== undefined) request.temperature = params.temperature;
  if (params.maxTokens !== undefined) request.max_tokens = params.maxTokens;
  if (params.topP !== undefined) request.top_p = params.topP;
  if (params.stop && params.stop.length > 0) request.stop = params.stop;

  return request;
}

/**
 * Parse an OpenAI completions response into an LLMResult.
 */
export function parseCompletionResponse(
  response: OpenAICompletionResponse,
  modelName: string
): LLMResult {
  const generationInfo: KServeGenerationInfo = {
    modelName: response.model ?? modelName,
    protocol: "openai",
  };

  if (response.usage) {
    generationInfo.tokenUsage = {
      promptTokens: response.usage.prompt_tokens,
      completionTokens: response.usage.completion_tokens,
      totalTokens: response.usage.total_tokens,
    };
  }

  return {
    generations: [
      response.choices.map((choice) => ({
        text: choice.text,
        generationInfo: {
          ...generationInfo,
          finishReason: choice.finish_reason,
        },
      })),
    ],
    llmOutput: generationInfo.tokenUsage
      ? { tokenUsage: generationInfo.tokenUsage }
      : undefined,
  };
}

/**
 * Parse a streaming OpenAI completion chunk.
 * Returns the text delta and optional usage, or null if nothing to emit.
 */
export function parseCompletionStreamChunk(
  raw: string
): { text: string; usage?: OpenAIUsage } | null {
  let chunk: OpenAICompletionStreamChunk;
  try {
    chunk = JSON.parse(raw) as OpenAICompletionStreamChunk;
  } catch {
    return null;
  }

  const text = chunk.choices?.[0]?.text;
  if (text === undefined || text === null) {
    // Could be a usage-only final chunk (empty choices)
    if (chunk.usage) {
      return { text: "", usage: chunk.usage };
    }
    return null;
  }
  return { text, usage: chunk.usage };
}
